### PR TITLE
add: Clear yum versionlock

### PR DESCRIPTION
**Pull request by fellipeh**
Users may have locked some of the installed packages to a certain version using `yum-plugin-versionlock`. This would most probably cause the conversion to fail because the RHEL equivalent of the package may easily have a different version.

This PR check if `yum-plugin-versionlock` is installed, if so, ask user if want to continue, if yes clear the versionlock list, if no conversion will stop.

See https://access.redhat.com/solutions/98873 for more information.

Signed-off-by: Fellipe Henrique <fpedrosa@redhat.com>

### CODE REVIEW

File: convert2rhel/pkghandler.py
Comment (bocekm) [reply convert2rhel/pkghandler.py:520]: Still to be changed..

**Review by bocekm — CHANGES_REQUESTED** [review]
A rollback of the version lock is needed.
Backing up and then copying back the `_VERSIONLOCK_FILE` is enough to do that.

**Comment by fellipeh** [comment]
Rebased

File: convert2rhel/unit_tests/pkghandler_test.py
Comment (SpyTec) [inline convert2rhel/unit_tests/pkghandler_test.py:99]: Add an assertion for the logger as well

File: convert2rhel/unit_tests/pkghandler_test.py
Comment (SpyTec) [inline convert2rhel/unit_tests/pkghandler_test.py:110]: Ideally extend `CallYumCmdMocked` to store the command called, and then assert that we call `yum versionlock clear`

**Review by SpyTec — CHANGES_REQUESTED** [review]
Missing tests for making sure backup works. As well as a function for restoring the backup and relevant tests

File: convert2rhel/pkghandler.py
Comment (fellipeh) [reply convert2rhel/pkghandler.py:517]: backup file already handle with this, here: https://github.com/oamg/convert2rhel/blob/master/convert2rhel/utils.py#L385

File: convert2rhel/pkghandler.py
Comment (fellipeh) [reply convert2rhel/pkghandler.py:531]: Backup this file was requested by @bocekm 

Regarding to restore:   agree here is not better place, but, rollback isn't the better place,because it's called once we cancel the process.. and the goal here is:  backup the original lock file, do all process and restore the original file.

File: convert2rhel/unit_tests/pkghandler_test.py
Comment (fellipeh) [reply convert2rhel/unit_tests/pkghandler_test.py:116]: Could you please elaborate more?

File: convert2rhel/pkghandler.py
Comment (fellipeh) [reply convert2rhel/pkghandler.py:523]: That's not truth.. version locks will not be cleared, it's will be cleared during the process, and it's supposed to restore to the original file after the process... that's the reason for backup/restore the file.

File: convert2rhel/pkghandler.py
Comment (SpyTec) [reply convert2rhel/pkghandler.py:531]: The goal here is to backup the lock file, and if conversion fails we roll it back. Restoring the file post migration will presumably have invalid entries due to the switch of packages to redhat repo, @bocekm do correct me if I'm wrong about the invalid entries.

@fellipeh, what you're doing here is backing up the file, clearing the file, then restoring it - all in the same function. This means the lock file is not changed and the function has no side effects except for touching the file. `yum versionlock clear` only clears the file, it does not do anything else. Yum itself checks this file whenever it handles packages whether a package it's handling is locked or not.

File: convert2rhel/unit_tests/pkghandler_test.py
Comment (SpyTec) [reply convert2rhel/unit_tests/pkghandler_test.py:116]: You need to assert that `call_yum_cmd()` was never called here.

Having multiple assertions avoids most false positives and ensures that the application runs as you expect it to.

For instance, ask yourself: would the tests pass if `call_yum_cmd()` was called before `ask_to_continue()`? If it would, then the tests are incorrect

File: convert2rhel/pkghandler.py
Comment (SpyTec) [inline convert2rhel/pkghandler.py:539]: I wasn't too keen on this sentence as it was using passive voice when asking the user for confirmation. I was struggling to come up with an active voice that worked, but Zack Hawkins helped and came up with this

```suggestion
		loggerinst.info("Do you want to continue? This will clear package version locks.")
```

File: convert2rhel/unit_tests/pkghandler_test.py
Comment (SpyTec) [inline convert2rhel/unit_tests/pkghandler_test.py:22]: Alphabetical import order
```suggestion
import sys
import yum
```

**Review by SpyTec — CHANGES_REQUESTED** [review]
A final round of changes, once it's all done I think this is good to merge

File: convert2rhel/unit_tests/pkghandler_test.py
Comment (SpyTec) [inline convert2rhel/unit_tests/pkghandler_test.py:89]: Can also refactor this to be `SysExitCallableObject`

File: convert2rhel/pkghandler.py
Comment (bocekm) [reply convert2rhel/pkghandler.py]: Yeah, the file should not be restored after the conversion. Since every package reinstalled during the conversions, there's a huge chance the restored version lock would not be valid anymore.
So, as Eric wrote, restore it only during the rollback.

**Comment by fellipeh** [comment]
@bocekm @SpyTec I believe this PR is ready to merge, right?

File: convert2rhel/pkghandler.py
Comment (bocekm) [inline convert2rhel/pkghandler.py:538]: I know that this has been changed two times already but the question _Do you want to continue?_ is kind of duplicate because `ask_to_continue()` prints _Continue with the system conversion? [y/n]_.
I actually liked the previous _"If you choose to continue, the package version locks will be cleared."_. But since you don't want to use the passive form, @SpyTec, what about:
```suggestion
        loggerinst.info("Upon continuing, we will clear all package version locks.")
```
I'm not sure if "we" is the best choice, but it's hard to choose something here when not using the passive form.

File: convert2rhel/pkghandler.py
Comment (bocekm) [inline convert2rhel/pkghandler.py:32]:
```suggestion
versionlock_file = utils.RestorableFile(_VERSIONLOCK_FILE_PATH)  # pylint: disable=C0103
```
To be consistent in the project, use lowercase as is the case of all the other RestorableFile instances (e.g. `system_release_file` or `yum_conf`).

**Review by bocekm — CHANGES_REQUESTED** [review]

File: convert2rhel/main.py
Comment (bocekm) [inline convert2rhel/main.py:64]: Please move this below the `subscription.rhn_reg_file.backup()`. Doing it as a first thing, before even checking if the system is possible to convert (gathering system info), is not right.

File: convert2rhel/pkghandler.py
Comment (bocekm) [inline convert2rhel/pkghandler.py:535]: Move the backup below the `utils.ask_to_continue()`. It makes no sense to back the file up in case they don't want to continue.
As you can see from the output, the information about the backup just adds a noise around the important message under it:
```
[10/06/2020 14:04:37] TASK - [Prepare: Clear yum version locks] *********************************
Backing up /etc/yum/pluginconf.d/versionlock.list
Copying /etc/yum/pluginconf.d/versionlock.list to /var/lib/convert2rhel/
WARNING - yum versionlock plugin is in use. It may cause the conversion to fail.
Do you want to continue? This will clear package version locks.

Continue with the system conversion? [y/n]: y
```

**Review by bocekm — CHANGES_REQUESTED** [review]

File: convert2rhel/pkghandler.py
Comment (SpyTec) [reply convert2rhel/pkghandler.py]: You're right, I thought it was just printing out `[y/n]` so we could be flexible. But yeah, that change SGTM. We should be using active voice as much as possible as it conveys better and sets a more direct tone

**Review by bocekm — APPROVED** [review]

**Comment by bocekm** [comment]
I'm merging this despite the one test failing. It seems to be a testing farm issue which I've filed at https://gitlab.com/testing-farm/cruncher/-/issues/11.

**Comment by bocekm** [comment]
I can't merge it before @SpyTec approves it.

**Review by SpyTec — APPROVED** [review]

**Pull request merged**
